cpp/src/tsne/barnes_hut.cuh

```diff
@@ -70,72 +71,75 @@ void Barnes_Hut(float *VAL, const int *COL, const int *ROW, const int NNZ,
+  MLCommon::device_buffer<float> radiusd(d_alloc, stream, 1);
+
+  TSNE::InitializationKernel<<<1, 1, 0, stream>>>(/*errl.data(),*/
+                                                  limiter.data(),
+                                                  maxdepthd.data(),
+                                                  radiusd.data());
   CUDA_CHECK(cudaPeekAtLastError());
 
   const int FOUR_NNODES = 4 * nnodes;
   const int FOUR_N = 4 * n;
   const float theta_squared = theta * theta;
   const int NNODES = nnodes;
 
-  // Actual mallocs
-  int *startl = (int *)d_alloc->allocate(sizeof(int) * (nnodes + 1), stream);
-  int *childl =
-    (int *)d_alloc->allocate(sizeof(int) * (nnodes + 1) * 4, stream);
-  float *massl =
-    (float *)d_alloc->allocate(sizeof(float) * (nnodes + 1), stream);
-  thrust::device_ptr<float> begin_massl = thrust::device_pointer_cast(massl);
+  // Actual allocations
+  MLCommon::device_buffer<int> startl(d_alloc, stream, nnodes + 1);
+  MLCommon::device_buffer<int> childl(d_alloc, stream, (nnodes + 1) * 4);
+  MLCommon::device_buffer<float> massl(d_alloc, stream, nnodes + 1);
+
+  thrust::device_ptr<float> begin_massl =
+    thrust::device_pointer_cast(massl.data());
   thrust::fill(thrust::cuda::par.on(stream), begin_massl,
                begin_massl + (nnodes + 1), 1.0f);
 
-  float *maxxl =
-    (float *)d_alloc->allocate(sizeof(float) * blocks * FACTOR1, stream);
-  float *maxyl =
-    (float *)d_alloc->allocate(sizeof(float) * blocks * FACTOR1, stream);
-  float *minxl =
-    (float *)d_alloc->allocate(sizeof(float) * blocks * FACTOR1, stream);
-  float *minyl =
-    (float *)d_alloc->allocate(sizeof(float) * blocks * FACTOR1, stream);
+  MLCommon::device_buffer<float> maxxl(d_alloc, stream, blocks * FACTOR1);
+  MLCommon::device_buffer<float> maxyl(d_alloc, stream, blocks * FACTOR1);
+  MLCommon::device_buffer<float> minxl(d_alloc, stream, blocks * FACTOR1);
+  MLCommon::device_buffer<float> minyl(d_alloc, stream, blocks * FACTOR1);
 
   // SummarizationKernel
-  int *countl = (int *)d_alloc->allocate(sizeof(int) * (nnodes + 1), stream);
+  MLCommon::device_buffer<int> countl(d_alloc, stream, nnodes + 1);
 
   // SortKernel
-  int *sortl = (int *)d_alloc->allocate(sizeof(int) * (nnodes + 1), stream);
+  MLCommon::device_buffer<int> sortl(d_alloc, stream, nnodes + 1);
 
   // RepulsionKernel
-  float *rep_forces =
-    (float *)d_alloc->allocate(sizeof(float) * (nnodes + 1) * 2, stream);
-  float *attr_forces = (float *)d_alloc->allocate(
-    sizeof(float) * n * 2, stream);  // n*2 double for reduction sum
+  MLCommon::device_buffer<float> rep_forces(d_alloc, stream, (nnodes + 1) * 2);
+  float *const rep_forces_c1 = rep_forces.data();
+  float *const rep_forces_c2 = rep_forces.data() + nnodes + 1;
+  MLCommon::device_buffer<float> attr_forces(
+    d_alloc, stream, n * 2);  // n*2 double for reduction sum
+  float *const attr_forces_c1 = attr_forces.data();
```

Comment:
Creating these convenience pointers makes the code more readable. Any reason for the pointer constness here?

Reply:
> Creating these convenience pointers makes the code more readable. 

Meaning I should add pointers consistently so that the function calls don't need `.data()`? I was mostly avoiding pointers because `buffer_base` implements a `resize` method that would invalidate pointers, but I did add them in a few places to dedupe logic.

Are you open to what I mentioned in my OP - adding `operator value_type*() { return _data; }`? That would be the best of both worlds: concise and pointers can't become invalid.

> Any reason for the pointer constness here?

Just so they can't be modified inadvertently. Let me know if you'd rather them be removed.

Reply:
@zbjornson, 

I tend to prefer the explicit `.data()` calls because they make it immediately obvious that these are pointers coming from device buffers. If there's no explicit reason to define extra pointers for this, my preference would be to just use the `.data()` calls in place (unless the pointer arithmetic becomes very hard to read). 

Would you be interested in opening up a new Github issue to propose the implicit conversion idea? This would allow others to discuss and provide feedback.

Reply:
@cjnolet Think I got it, but to be totally clear, you're suggesting removing `float *const rep_forces_c2` and just use `rep_forces.data() + nnodes + 1` in the function call? Or leave it as-is, or remove the `const`, or add e.g. `int const* sortl_ = sortl.data()`?

I can open an issue to discuss implicit conversion.

Reply:
I'd prefer just to perform the pointer arithmetic inline here since it's not all that complicated